docs/step-composition.md

```diff
@@ -0,0 +1,203 @@
+
+#### GitHub URIs
+
+The most common place of reusing tasks from git will be [GitHub](https://github.com/) which is where most open source software lives including the [Tekton Catalog](https://github.com/tektoncd/catalog) so to keep the Tekton YAML concise we support a github specific notation of:
+
+```
+ownerOrUser/repositoryName/pathToResource[@branchTagOrSHA]
+```
+                                                                                              
+e.g. to reference a Tekton catalog task use: [tektoncd/catalog/task/kaniko/0.1/kaniko.yaml](https://github.com/tektoncd/catalog/task/kaniko/0.1/kaniko.yaml) which will use the latest version of `0.1` of the `kaniko` task
+
+#### Other git servers
+
+For other git servers we support the following syntax where the `gitCloneURI` ends with `.git`
+
+```
+gitCloneURI/pathToResource[@branchTagOrSHA]
+```
+                     
+So the following examples can be used:
+    
+* `https://github.com/myowner/myrepo.git/some/path.yaml`
+* `git@github.com:bar/foo.git/thingy.yaml@mybranch`  
+* `https://mybitbucket.com/scm/myowner/myrepo.git/thingy.yaml@mybranch`
+* `git://host.xz/org/repo.git/some/path.yaml@v1.2.3`
+
+### kubernetes resources
+
+To reference a `Task` in kubernetes use the usual `TaskRef` notation...
+
+```yaml
+- uses: 
+    ref: 
+      name: my-task
+```
+
+### OCI bundles
+
+To reference a `Task` in an OCI bundle use the usual `TaskRef` notation...
+
+```yaml
+- uses: 
+    ref: 
+      bundle: gcr.io/myproject/mybundle:1.2.3
+      name: my-task
+```
+
+
+
+## Reusing all steps of a Task
+
+To reuse all the steps of a Task inside your task add a `uses:`  to a step [like this example](../pkg/stepper/test_data/tests/task/input.yaml) 
+
+```yaml 
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  generateName: my-pipeline-
+spec:
+  steps:
+  - uses:
+      git: tektoncd/catalog/task/kaniko/0.1/kaniko.yaml
+  - name: my-custom-step
+    image: golang:1.15
+    script: |
+      script: |
+        #!/bin/sh
+        echo "hello world"
+```
+
+This will then reuse all of the steps in the [tektoncd/catalog/task/kaniko/0.1/kaniko.yaml](https://github.com/tektoncd/catalog/task/kaniko/0.1/kaniko.yaml) file in place inside your `Task` before the **my-custom-step**
+
+The [generated Task is here](../pkg/stepper/test_data/tests/task/expected.yaml)
+
+### Customising all reused steps
+
+You may want to add some common environment variables or volume mounts to the used steps. You can do this by adding the `env`, `envFrom` or `volumeMount` properties to your step [like this example](../pkg/stepper/test_data/tests/task_override_all_steps/input.yaml)
+
+e.g. lets add a **MY_VAR** environment variable to all reused steps:
+
+
+```yaml 
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  generateName: my-pipeline-
+spec:
+  steps:
+  - uses:
+      git: tektoncd/catalog/task/kaniko/0.1/kaniko.yaml
+    env:
+      - name: MY_VAR
+        value: CHEESE
+  - name: my-custom-step
```

Comment:
The `env` (and similar) customisation feels slightly odd to me at this hierarchy level.

It reminds me of https://tekton.dev/docs/pipelines/tasks/#specifying-a-step-template, with the difference that the `stepTemplate` is a default, and here we have an override. What about making this explicit, and wrapping it e.g. `stepOverride`? Like this:

```yaml
spec:
   steps:
   - uses:
       git: tektoncd/catalog/task/kaniko/0.1/kaniko.yaml
     stepOverride:
       env:
       - name: MY_VAR
         value: CHEESE
```

Reply:
having a Container struct nested in a Container for one container feels odd to me too ;).

We've found its often handy to start off reusing a step; [extending it, then even completely inlining it](https://jenkins-x.io/v3/develop/pipelines/catalog/#overriding-a-pipeline-step-locally) (so that its no longer extending another step) and then going back to extending again. 

During this change, it feels clumsy to have to keep wrapping the environment variables/commands/volume properties in a `stepOverride:` block and then unwrap again. Also it's going to cause confusion that in a step there are 2 places you can put `env:` and `script:` etc and developers have to remember which place to put things and deal with errors when folks forget to wrap/unwrap. I'm not sure all that complexity and confusion helps make things more clear.

e.g. think of Java for a second and a class. When you extend a class...

```java
class Cheese extends Wine {
  // members
}
```
at any time you can add or remove the `extends Wine` and by adding/removing the extension you don't have to go around wrapping/unwrapping the members - the extension clause is enough and folks don't get confused. 

So if you want to make the override mode more clear, why don't we use `override` instead of `uses` / `ref`?
```yaml
spec:
   steps:
   - override:
       git: tektoncd/catalog/task/kaniko/0.1/kaniko.yaml
       step: my-step
     env:
     - name: MY_VAR
       value: CHEESE
```

we don't need to say `stepOverride` as we are in a step so its clear we are overriding a step. Then, rather like Java, we can add/remove the `override` at any time without having to make large changes to the rest of the step and the `override` mode is clear?